### ISSUE TITLE
REST API doesn't correctly return relations that are not inversed

### ISSUE THREAD

**Issue by laurenskling** [strapi/strapi#21355]
### Required System information

- Node.js version: 20
- NPM version: yarn 1.22
- Strapi version: 5.0.0 (and rc 17-29 at least)
- Database: sqlite
- Operating system: macos
- Is your project Javascript or Typescript: js

### Describe the bug + Steps to reproduce the behavior

I have a Homepage singleType. It doesnt have draftAndPublish, because the homepage never goes unpublished (v4 logic of DP). 

Now say i have a form contentType i want to relate. I pick it in the UI. I set all permission. I go to:
`http://0.0.0.0:1337/api/homepage/homepage?&populate[form][fields]=*` (/homepage/homepage is my custom endpoint), i see `form: null`.
if I then request: `http://0.0.0.0:1337/api/homepage/homepage?&populate[form][fields]=*&status=draft`, my form is returned. This is weird. I believe that without setting status, it should always work. I think the system assumes it will bubble down from the top contentType (in this case homepage), but that doesn't have a status.

### Expected behavior

A non-DP contentType should always return all of the requested relations.

### Status

- we found it has to do something with relations without `mappedBy` and `inversedBy` fields
- probably nothing to do with non-DP

**Comment by alexandrebodin** [comment]
Hi, 

Could you provide the content-type schema.json for reproduction ?

I'm sorry but you lost me on the component explanation. Do you mind recording a loom or writing a list of specific steps to reproduce 🙏  ? 

A few infos

- `production` is not a valid status it's either `draft` or `published`
- You should see the published versions of your relations (when you target a D&P content-type) from your homepage when you populate. We are setting the status as published by default unless you manually overwrite it (might be where there is a bug)

**Comment by laurenskling** [comment]
Yeah im sorry about the component explanation. It's very specific. It took me a while to understand what was happening. I'll try to make a repro a.s.a.p.

turns out, every &status=x just works (probably draft). So I can change my story to just `draft` returning my relation, whilst nothing or `published` doesn't.

**Comment by laurenskling** [comment]
Okay I got it about the direct relation being `null`. (and probably also why components return `null`).

I've created a one way relation without `mappedBy` and `inversedBy` on the other side.

Steps:
1. npx create-strapi my-strapi-project with the test data installed
2. in the article schema.json, remove the author block
3. in the author schema.json, remove the `mappedBy`
4. start the app
5. make articles
6. make an author and relate some articles
7. set permissions to public rest endpoints
8. http://0.0.0.0:1337/api/authors?populate[articles][fields]=*
9. 
```
// http://0.0.0.0:1337/api/authors?populate[articles][fields]=*

{
  "data": [
    {
      "id": 1,
      "documentId": "fww35rioayia8yk9s9jzzs7g",
      "name": "Henk",
      "email": null,
      "createdAt": "2024-09-18T21:32:27.022Z",
      "updatedAt": "2024-09-18T21:33:11.418Z",
      "publishedAt": "2024-09-18T21:33:11.414Z",
      "locale": null,
      "articles": [
        
      ]
    }
  ],
  "meta": {
    "pagination": {
      "page": 1,
      "pageSize": 25,
      "pageCount": 1,
      "total": 1
    }
  }
}
```
<img width="1470" alt="Scherm­afbeelding 2024-09-18 om 23 38 18" src="https://github.com/user-attachments/assets/71783e83-c366-4fc2-b269-57e5724d389f">

**Comment by laurenskling** [comment]
```
// http://0.0.0.0:1337/api/authors?populate[articles][fields]=*&status=draft

{
  "data": [
    {
      "id": 1,
      "documentId": "fww35rioayia8yk9s9jzzs7g",
      "name": "Henk",
      "email": null,
      "createdAt": "2024-09-18T21:32:27.022Z",
      "updatedAt": "2024-09-18T21:41:34.948Z",
      "publishedAt": "2024-09-18T21:41:34.942Z",
      "locale": null,
      "articles": [
        {
          "id": 1,
          "documentId": "ef3ievdpz7e5zvuaf8eodug9",
          "title": "first article",
          "description": null,
          "slug": null,
          "createdAt": "2024-09-18T21:30:17.611Z",
          "updatedAt": "2024-09-18T21:42:04.948Z",
          "publishedAt": null,
          "locale": null
        },
        {
          "id": 5,
          "documentId": "ufn2vyouiuzyd2uzdi0q2czc",
          "title": "second article in draft",
          "description": null,
          "slug": null,
          "createdAt": "2024-09-18T21:33:06.971Z",
          "updatedAt": "2024-09-18T21:42:18.291Z",
          "publishedAt": null,
          "locale": null
        }
      ]
    }
  ],
  "meta": {
    "pagination": {
      "page": 1,
      "pageSize": 25,
      "pageCount": 1,
      "total": 1
    }
  }
}
```

in draft, it shows up.
I wasn't totally crazy haha.

**Comment by alexandrebodin** [comment]
@laurenskling If you are targeting a content-type that has D&P enabled you won't see the data until it's actually published even if you link to it from a non D&P content-type. This is normal to avoid leaking any data.

Maybe be worth splitting this into 2 different issues so we can get them sorted out more clearly 🙏

**Comment by laurenskling** [comment]
The relations are published. You can see it on the screenshot. If you
follow the steps, you can reproduce it with the test data installer

Will split off the component issue soon. Let's focus this one on the one
way relation not populating.

Op do 19 sep. 2024 14:05 schreef Alexandre BODIN ***@***.***>:

**Comment by alexandrebodin** [comment]
@laurenskling Have you noticed if it's only happening on single-types by any chance ?

**Comment by laurenskling** [comment]
My original discovery was with a single type. The reproduction steps, its with 2 collection types

**Comment by laurenskling** [comment]
Awesome, thank you <3

**Comment by laurenskling** [comment]
Split: https://github.com/strapi/strapi/issues/21382

**Comment by laurenskling** [comment]
Just tested: it doesn't even have something to do with being non-DP. Its the not inversed that is probably the main reason

**Comment by alexandrebodin** [comment]
So It's actually not a bug with the fetching but the relation being unidirectionnal and not being copied over when you publish the targted content-type. Pretty tricky but the data has a pbl. I'll bump this as critical and notif the team 🙏

**Comment by laurenskling** [comment]
I don't know what a `pbl` is, but I'm glad we are getting closer :)

**Comment by alexandrebodin** [comment]
> I don't know what a `pbl` is, but I'm glad we are getting closer :)

problem :D sorry !

**Comment by alexandrebodin** [comment]
@laurenskling Ok this is a very complex fix to implement and we won't be able to do it today. Hopefully we will get it done as early as possible next week. Thank you for your help!

**Comment by alexandrebodin** [comment]
Hey @laurenskling I made an experimental fix here 0.0.0-experimental.fed75ee8e64c57dbed0b670b25ef026b69baab10

Could you try it out and tell me if that fixes it for you 🙏  

FYI you might need to re-add your relation (unselect / save / select again) to cleanup the data.

**Comment by laurenskling** [comment]
Almost :)

When the source is non-DP, it works in published, but not in draft (this sounds wrong to do, but I see myself calling my non-DP homepage with status=draft to see my draft sections on it)

1. https://github.com/laurenskling/strapi-5-tests/tree/main
2. add an article (this is DP)
3. add an author (this in non-DP) and relate it to the article
4. http://0.0.0.0:1337/api/authors?populate[articles][fields]=*&status=published  had articles
5. http://0.0.0.0:1337/api/authors?populate[articles][fields]=*&status=draft does NOT have articles

(relates to https://github.com/strapi/strapi/issues/21382)

**Comment by laurenskling** [comment]
(when i change author to DP, it works without changing the data first, so it seems to not be related to connecting stuff on save)

**Comment by alexandrebodin** [comment]
> Almost :)
> 
> When the source is non-DP, it works in published, but not in draft (this sounds wrong to do, but I see myself calling my non-DP homepage with status=draft to see my draft sections on it)
> 

This is expected. Even if your source doesn't have DP it will still load published only by default to avoid leaking private data. and you need the status=draft option to load the draft data.

> 1. [laurenskling/strapi-5-tests@`main`](https://github.com/laurenskling/strapi-5-tests/tree/main?rgh-link-date=2024-09-20T06%3A05%3A50Z)
> 2. add an article (this is DP)
> 3. add an author (this in non-DP) and relate it to the article
> 4. [http://0.0.0.0:1337/api/authors?populate[articles][fields]=*&status=published](http://0.0.0.0:1337/api/authors?populate%5Barticles%5D%5Bfields%5D=*&status=published)  had articles
> 5. [http://0.0.0.0:1337/api/authors?populate[articles][fields]=*&status=draft](http://0.0.0.0:1337/api/authors?populate%5Barticles%5D%5Bfields%5D=*&status=draft) does NOT have articles
> 
> (relates to #21382)

**Comment by laurenskling** [comment]
It's not that it loads published by default and not drafts, _I am unable to loads drafts at all_.

Fetching my non-DP content type with `status=draft` will _not_ load DP relations. Published/modified or draft. Nothing.

If you follow my steps you can reproduce

**Comment by laurenskling** [comment]
yes

**Comment by alexandrebodin** [comment]
Ok thank you that's another issue I didn't notice 😭

**Comment by laurenskling** [comment]
@alexandrebodin  on 0.0.0-experimental.fed75ee8e64c57dbed0b670b25ef026b69baab10
is now the same situation as https://github.com/strapi/strapi/issues/21355#issuecomment-2363702776

working, except for non-DP requesting the DP relation drafts (as configed the repo for this issue)

**Comment by alexandrebodin** [comment]
Hey @laurenskling, I changed the fix and made a new PR. I believe this is addressing both the content-type & component relation issues you found. 

If you can give it a look 🙏  `0.0.0-experimental.727a0f9b729da91886ca93e88ec1d6d811d1beba` 

_you will need to recreate your relations_

**Comment by laurenskling** [comment]
Initial impression is very good. Both use cases (this and https://github.com/strapi/strapi/issues/21382) seem to be working.
only thing I can find right now, is that after migration, current use case and status=draft give me `null`. After recreating the relation, it works. But you can imagine I'd rather have it working from the migration :) Open a new ticket for it?

**Comment by alexandrebodin** [comment]
@laurenskling do you mean migration from v4 to v5 or between your current v5 to the experimental ?

**Comment by laurenskling** [comment]
@alexandrebodin yeah a v4 db. Whenever I test this I take my original v4 db again. Thats pretty easy with sqlite :)

**Comment by alexandrebodin** [comment]
Ok I didn't think about a migration issue for this during the fix. I'll ping the team to create a ticket for it 👌

**Comment by laurenskling** [comment]
Thanks :) you don't need me to write about it? Can you someway include me in an alert when it gets fixed so I can test it?

**Comment by alexandrebodin** [comment]
@laurenskling If you want to follow it I think the best is your create an issue in that case 👍 It's probably best to give visibility actually 🙏

**Comment by alexandrebodin** [comment]
@laurenskling FYI I updated the PR, I believe it fixes the migration issue you were having. I made a new experimental here

`0.0.0-experimental.9bd63a0f51283779d88aa9d86aa78c40ed628a15`

**Comment by alexandrebodin** [comment]
New experimental :) `0.0.0-experimental.f8a68bc03b751aa0a66c4dcfaf83553c9b6adf2b`

**Comment by laurenskling** [comment]
this:
<img width="1412" alt="Scherm­afbeelding 2024-09-24 om 15 04 02" src="https://github.com/user-attachments/assets/f6fc7915-0d38-4e9a-aee7-7dcf38139cb9">

and `null` in the REST api. :(

**Comment by laurenskling** [comment]
> @laurenskling FYI I updated the PR, I believe it fixes the migration issue you were having. I made a new experimental here
> 
> `0.0.0-experimental.9bd63a0f51283779d88aa9d86aa78c40ed628a15`

this one did not have the double entry in the UI but still `null` in the REST API.

**Comment by alexandrebodin** [comment]
Ok I'll investigate this too

**Comment by laurenskling** [comment]
yes, this is a non-DP Homepage, having a repeatable component, that contain a manyToOne relation (without inverse) to DP sections. I don't use i18n in my v4 db.

**Comment by laurenskling** [comment]
> Initial impression is very good. Both use cases (this and #21382) seem to be working. only thing I can find right now, is that after migration, current use case and status=draft give me `null`. After recreating the relation, it works. But you can imagine I'd rather have it working from the migration :) Open a new issue for it?

here i said current use case, but thats not true. its the https://github.com/strapi/strapi/issues/21382 use case.

**Comment by alexandrebodin** [comment]
Ok cool I'll replicate that 👍

**Comment by laurenskling** [comment]
on homepage:

```
sections: {
  type: 'component',
  repeatable: true,
  component: 'pagebuilder.section',
},
```

component:
```
module.exports = {
  collectionName: 'components_pagebuilder_section',
  info: {
    displayName: 'Section',
    singularName: 'section-component',
    pluralName: 'section-components',
    description: '',
    icon: 'grid',
  },
  options: {},
  attributes: {
    section: {
      type: 'relation',
      relation: 'manyToOne',
      target: 'plugin::pagebuilder.section',
    },
  },
};
```

(stripped down) section content type:
```
module.exports = {
  kind: 'collectionType',
  info: {
    singularName: 'section', // kebab-case mandatory
    pluralName: 'sections', // kebab-case mandatory
    displayName: 'Pagebuilder section',
    description: 'A section containing columns or other dynamic content',
  },
  options: {
    draftAndPublish: true,
    privateAttributes: ['createdAt'],
  },
  pluginOptions: {
    'content-manager': {
      visible: true,
    },
    'content-type-builder': {
      visible: false,
    },
  },
  attributes: {
    name: {
      type: 'string',
      required: true,
    },
    components: {
      type: 'dynamiczone',
      components: [
        'pagebuilder.column',

      ],
    },
  },
};
```

**Comment by alexandrebodin** [comment]
@laurenskling how did you create that relation ? you are not supposed to be able to have manyToOne inside components at all 😬

**Comment by laurenskling** [comment]
hahaha hackerman! I never use the Content Builder, I write my own schema's in JS.

**Comment by laurenskling** [comment]
half my codebase is relations in components in DZ's 🤷 how are you not supposed to do that 😅

**Comment by laurenskling** [comment]
the code above is a relic from when I wasnt able to reorder relations. But still, things like a button component in a DZ that links to a page, you have relation to the page, right? It's a super common use case.

**Comment by alexandrebodin** [comment]
Yes yes your usecase is valid and you can do it with the unidirection oneToOne. Because it's unidirectional it's actually just a manyToOne in DB. 

In the context of components only undirectional relations make sense as you never share a component data across multiple entries it's always standalone. And we only support oneToOne and oneToMany which are equivalent to manyToOne & manyToMany in the non unidirectional world.

I think it's just working because it's using the same storage structure but I reco ou switch to oneToOne in your case and it will work the same way but be valid in the content-type-builder

📓 We should definitly add a warning when we detect these

**Comment by laurenskling** [comment]
Ah, so it's just a misunderstanding of me of how to use the different relations. It's kind of hard to wrap your head around if your not a database type of person. I always try to think of it as (could be used multiple times)To(single or multiple). I never thought about the unidirection because I don't understand it ;) 
So If i would change in the component:
```
    section: {
      type: 'relation',
      relation: 'manyToOne',
      target: 'plugin::pagebuilder.section',
    },
```
to
```
  section: {
      type: 'relation',
      relation: 'oneToOne',
      target: 'plugin::pagebuilder.section',
    },
```

it would work the same for me but be correct code, yes?

**Comment by laurenskling** [comment]
will give it a spin a.s.a.p. :) thanks <3

**Comment by alexandrebodin** [comment]
Running a new experimental right there to fix the multiple relations appear in the UI https://github.com/strapi/strapi/actions/runs/11016378985

**Comment by laurenskling** [comment]
The multiple relation in the UI bug is gone :)

**Comment by laurenskling** [comment]
Sadly, I did not get my `null` relations fixed by setting oneToOne. Even setting them to oneToOne first on v4.

**Comment by alexandrebodin** [comment]
I fixed a few more things last night. I'll make a new experimental this morning. I cannot reproduce the null in the api. Have you check your permissions are all configure to access all the content-types ?

**Comment by laurenskling** [comment]
Yes, because when i change the relations in the UI to something else, the draft works. So it's only after migration.

**Comment by alexandrebodin** [comment]
Ok so just to confirm

- Right after migration you get null data in the API
- If you unselect / reselect in the UI -> you see the data in the API correctly ?

**Comment by alexandrebodin** [comment]
Ok 🤞 @laurenskling  This should fix the migration and the post migration relations 

https://github.com/strapi/strapi/actions/runs/11035755028/job/30652538862

**Comment by laurenskling** [comment]
@alexandrebodin I'm sorry mate, I've found a broken use case again, running 5.0.2.

Having my setup like: https://github.com/strapi/strapi/issues/21355#issuecomment-2371486655
(Changed the relation to `oneToOne`, doesn't change it)

Out of the box after migration it works. When I change a relation in the UI, it goes `null` again in the API :( 
When I go to the related entry (section), and publish changes there, it _does_ show up in my homepage API results. But it will be gone again when changing (or adding) a relation.

**Comment by laurenskling** [comment]
By the way, this is the https://github.com/strapi/strapi/issues/21382 use case again.

**Comment by Marc-Roig** [comment]
Hey @laurenskling , alex is coming back on wednesday and we are going to discuss a fix for  this. 
This is only happening on anyToOne relations , I believe the issue is we need to connect both the draft &  published version and its being trimmed to only one relation. I am reopening the issue and we will investigate

**Comment by dzakki** [comment]
Hi team, perhaps this #21623 is similar to this issue. could you please check it?

**Comment by Marc-Roig** [comment]
@laurenskling I have not been able to reproduce your issue, I have a oneToOne relation (one that is not inversed)
after a migration, if I change the relation and I target an entry thats published I see that relation in the api. Could you open another ticket so we can better track it and discuss it ?

**Comment by laurenskling** [comment]
Thanks for looking into it @Marc-Roig , there is: https://github.com/strapi/strapi/issues/21382

**Comment by Marc-Roig** [comment]
I have a strong suspicion https://github.com/strapi/strapi/pull/21745 might fix this issue, I assume you are using the upgrade tool to migrate to v5 ?

**Comment by laurenskling** [comment]
@Marc-Roig actually, I just start my project with v5 installed in the package.json with a v4 db, and all migrations run by themselves. I havent used the upgrade tool

**Comment by laurenskling** [comment]
Cannot reproduce it anymore as of 5.0.3!

**Comment by Marc-Roig** [comment]
weird! maybe that was an issue during the release and some packages were not correctly published 🤔 we will investigate! thank you for sharing it
